qiskit/algorithms/optimizers/gradient_descent.py

```diff
@@ -95,48 +145,176 @@ def grad_f(x):
+    and is more familiar with step by step
+    optimization algorithms, :meth:`~.step` has been created to acts as a wrapper for :meth:`~.ask`
+    and :meth:`~.tell`.
+    In the same spirit the method :meth:`~.minimize` will optimize the function and return the result
+    directly.
+
+    To see other libraries that use this interface one can visit:
+    https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html
 
-            print(f"Found minimum {x_opt} at a value of {fx_opt} using {nfevs} evaluations.")
 
     """
 
     def __init__(
         self,
         maxiter: int = 100,
-        learning_rate: Union[float, Callable[[], Iterator]] = 0.01,
+        learning_rate: Union[float, List[float], np.ndarray, Callable[[], Iterator]] = 0.01,
         tol: float = 1e-7,
         callback: Optional[CALLBACK] = None,
         perturbation: Optional[float] = None,
     ) -> None:
-        r"""
+        """
         Args:
             maxiter: The maximum number of iterations.
-            learning_rate: A constant or generator yielding learning rates for the parameter
-                updates. See the docstring for an example.
+            learning_rate: A constant, list, array or factory of generators yielding learning rates
+                           for the parameter updates. See the docstring for an example.
             tol: If the norm of the parameter update is smaller than this threshold, the
-                optimizer is converged.
-            perturbation: If no gradient is passed to ``GradientDescent.optimize`` the gradient is
-                approximated with a symmetric finite difference scheme with ``perturbation``
+                optimizer has converged.
+            perturbation: If no gradient is passed to :meth:`~.minimize` the gradient is
+                approximated with a forward finite difference scheme with ``perturbation``
                 perturbation in both directions (defaults to 1e-2 if required).
-                Ignored if a gradient callable is passed to ``GradientDescent.optimize``.
+                Ignored when we have an explicit function for the gradient.
+        Raises:
+            ValueError: If ``learning_rate`` is an array and its lenght is less than ``maxiter``.
         """
-        super().__init__()
-
-        self.maxiter = maxiter
-        self.learning_rate = learning_rate
-        self.perturbation = perturbation
-        self.tol = tol
+        super().__init__(maxiter=maxiter)
         self.callback = callback
+        self._state: Optional[GradientDescentState] = None
+        self._perturbation = perturbation
+        self._tol = tol
+        # if learning rate is an array, check it is sufficiently long.
+        if isinstance(learning_rate, (list, np.ndarray)):
+            if len(learning_rate) < maxiter:
+                raise ValueError(f"Length of learning_rate is smaller than maxiter ({maxiter}).")
+        self._learning_rate = learning_rate
+
+    @property
```

Comment:
From my POV you could drop all these trivial getters and setters, but you can of course keep them if you want

Reply:
We could put it in the class docstring. But if the documentation is just "this is the state" then maybe we don't neeneed it at all 😉

Reply:
Okay, I will change it in the next push.